From 62e53f2697948526bbbd8f6ad9e100e0fd53565d Mon Sep 17 00:00:00 2001
From: T S Rameshkumar <rameshkumar.t@phytecembedded.in>
Date: Mon, 12 Oct 2026 11:09:08 +0530
Subject: [PATCH 3/5] imx8mm-phyboard-polis: Map SD2_BOOT to MMC2 and annotate
 SPL DT nodes

Map SD2_BOOT to BOOT_DEVICE_MMC2 (usdhc2 SD slot) in SPL and annotate USDHC2, USDHC3, regulator, GPIO, and UART3 console nodes with bootph-pre-ram tags.
---
 .../dts/imx8mm-phyboard-polis-rdk-u-boot.dtsi |  8 +++
 board/phytec/phycore_imx8mm/spl.c             | 61 ++++++++++++++++++-
 2 files changed, 67 insertions(+), 2 deletions(-)

diff --git a/arch/arm/dts/imx8mm-phyboard-polis-rdk-u-boot.dtsi b/arch/arm/dts/imx8mm-phyboard-polis-rdk-u-boot.dtsi
index 516e52e1f5d..ef6da880dba 100644
--- a/arch/arm/dts/imx8mm-phyboard-polis-rdk-u-boot.dtsi
+++ b/arch/arm/dts/imx8mm-phyboard-polis-rdk-u-boot.dtsi
@@ -69,3 +69,11 @@
 &wdog1 {
 	bootph-pre-ram;
 };
+
+&reg_usdhc2_vmmc {
+	bootph-pre-ram;
+};
+
+&pinctrl_reg_usdhc2_vmmc {
+	bootph-pre-ram;
+};
diff --git a/board/phytec/phycore_imx8mm/spl.c b/board/phytec/phycore_imx8mm/spl.c
index 690a51f7a72..9983d9bb5c2 100644
--- a/board/phytec/phycore_imx8mm/spl.c
+++ b/board/phytec/phycore_imx8mm/spl.c
@@ -18,17 +18,29 @@
 #include <log.h>
 #include <spl.h>
 
+#include "lpddr4_timing.h"
+#include "../common/imx8m_som_detection.h"
+
+enum phytec_imx8mm_ddr_eeprom_code {
+	PHYTEC_IMX8MM_DDR_1GB = 1,
+	PHYTEC_IMX8MM_DDR_2GB = 2,
+	PHYTEC_IMX8MM_DDR_4GB = 3,
+};
+
 DECLARE_GLOBAL_DATA_PTR;
 
+#define EEPROM_ADDR		0x51
+#define EEPROM_ADDR_FALLBACK	0x59
+
 int spl_board_boot_device(enum boot_device boot_dev_spl)
 {
 	switch (boot_dev_spl) {
 	case SD2_BOOT:
 	case MMC2_BOOT:
-		return BOOT_DEVICE_MMC1;
+		return BOOT_DEVICE_MMC2;
 	case SD3_BOOT:
 	case MMC3_BOOT:
-		return BOOT_DEVICE_MMC2;
+		return BOOT_DEVICE_MMC2_2;
 	case QSPI_BOOT:
 		return BOOT_DEVICE_NOR;
 	case USB_BOOT:
@@ -40,6 +52,45 @@ int spl_board_boot_device(enum boot_device boot_dev_spl)
 
 static void spl_dram_init(void)
 {
+	int ret;
+	enum phytec_imx8mm_ddr_eeprom_code size = PHYTEC_EEPROM_INVAL;
+
+	ret = phytec_eeprom_data_setup_fallback(NULL, 0, EEPROM_ADDR,
+			EEPROM_ADDR_FALLBACK);
+	if (ret && !IS_ENABLED(CONFIG_PHYCORE_IMX8MM_RAM_SIZE_FIX))
+		goto out;
+
+	ret = phytec_imx8m_detect(NULL);
+	if (!ret)
+		phytec_print_som_info(NULL);
+
+	if (IS_ENABLED(CONFIG_PHYCORE_IMX8MM_RAM_SIZE_FIX)) {
+		if (IS_ENABLED(CONFIG_PHYCORE_IMX8MM_RAM_SIZE_1GB))
+			size = PHYTEC_IMX8MM_DDR_1GB;
+		else if (IS_ENABLED(CONFIG_PHYCORE_IMX8MM_RAM_SIZE_2GB))
+			size = PHYTEC_IMX8MM_DDR_2GB;
+		else if (IS_ENABLED(CONFIG_PHYCORE_IMX8MM_RAM_SIZE_4GB))
+			size = PHYTEC_IMX8MM_DDR_4GB;
+	} else {
+		size = phytec_get_imx8m_ddr_size(NULL);
+	}
+
+	switch (size) {
+	case PHYTEC_IMX8MM_DDR_1GB:
+		set_dram_timings_1gb();
+		break;
+	case PHYTEC_IMX8MM_DDR_2GB:
+		break;
+	case PHYTEC_IMX8MM_DDR_4GB:
+		set_dram_timings_4gb();
+		break;
+	default:
+		goto out;
+	}
+	ddr_init(&dram_timing);
+	return;
+out:
+	puts("Could not detect correct RAM size. Fall back to default.\n");
 	ddr_init(&dram_timing);
 }
 
@@ -48,6 +99,12 @@ int board_fit_config_name_match(const char *name)
 	return 0;
 }
 
+void spl_board_init(void)
+{
+	arch_misc_init();
+	puts("Normal Boot\n");
+}
+
 void board_init_f(ulong dummy)
 {
 	int ret;
-- 
2.34.1

